src/filters/sse.rs

```diff
@@ -52,119 +53,126 @@ use futures::{future, Stream, TryStream, TryStreamExt};
 use http::header::{HeaderValue, CACHE_CONTROL, CONTENT_TYPE};
 use hyper::Body;
 use pin_project::pin_project;
-use serde::Serialize;
-use serde_json;
+use serde_json::{self, Result as SerdeResult};
 use tokio::time::{self, Delay};
 
-use self::sealed::{
-    BoxedServerSentEvent, EitherServerSentEvent, SseError, SseField, SseFormat, SseWrapper,
-};
+use self::sealed::SseError;
 use super::header;
 use crate::filter::One;
 use crate::reply::Response;
 use crate::{Filter, Rejection, Reply};
 
-/// Server-sent event message
-pub trait ServerSentEvent: SseFormat + Sized + Send + Sync + 'static {
-    /// Convert to either A
-    fn into_a<B>(self) -> EitherServerSentEvent<Self, B> {
-        EitherServerSentEvent::A(self)
+// Server-sent event data type
+#[derive(Debug)]
+enum DataType {
+    Text(String),
+    Json(SerdeResult<String>),
+}
+
+/// Server-sent event
+#[derive(Default, Debug)]
+pub struct Event {
+    name: Option<String>,
+    id: Option<String>,
+    data: Option<DataType>,
+    event: Option<String>,
+    comment: Option<String>,
+    retry: Option<Duration>,
+}
+
+impl Event {
+    /// Set Server-sent event data
+    /// data field(s) ("data:<content>")
+    pub fn set_data<T: Display>(mut self, data: Option<T>) -> Event {
+        self.data = data.map(|d| DataType::Text(d.to_string()));
+        self
     }
 
-    /// Convert to either B
-    fn into_b<A>(self) -> EitherServerSentEvent<A, Self> {
-        EitherServerSentEvent::B(self)
+    /// Set Server-sent event data
+    /// data field(s) ("data:<content>")
+    pub fn set_json_data<T: Serialize>(mut self, data: Option<T>) -> Event {
+        self.data = match data {
+            Some(data) => Some(DataType::Json(serde_json::to_string(&data))),
+            None => None,
+        };
+        self
     }
 
-    /// Convert to boxed
-    fn boxed(self) -> BoxedServerSentEvent {
-        BoxedServerSentEvent(Box::new(self))
+    /// Set Server-sent event comment
+    /// Comment field (":<comment-text>")
+    pub fn set_comment<T: Display>(mut self, comment: Option<T>) -> Event {
```

Comment:
For method naming, I'd say that builders usually can skip the `set_` prefix. Most often, `set_foo` operates on `&mut self`.

What's the likelihood that we'd also want to add getters to `Event`? If `id()` should return a reference, then I'd say the builder methods should be prefixed `with_`. If we probably don't want getters, we can skip prefixes altogether.

I don't yet know which is more useful.

Reply:
yeah the previous api did not not have getters, i'd say we stick with only setters